### PR TITLE
first pass for configurable readaligner

### PR DESCRIPTION
**Pull request by mr-c**
For @ctb

### CODE REVIEW

**Comment by ctb** [comment]
Thanks! What are the plans moving forward?

Also, could you pull in #750? I'm happy to work on fixing the issues with that over the next few weeks and predicate this merge on that one, but it'd be nice to not have to do a lot of merging.

**Comment by mr-c** [comment]
I'd rather merge this in and update #750 to use this as it doesn't change the behavior.

To-do:

- [ ] fill out pydoc
- [ ] update the [`pairhmm_learning/learn.py`](https://github.com/ged-lab/2013-read-aligner/blob/mrc-additions/pairhmm_learning/learn.py) script to output compatible python code

**Comment by ctb** [comment]
Ok 

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by ctb** [comment]
I would also ask:

- [ ] provide file loading routines in Python
- [ ] provide several example sets of parameters, for low k and high k

**Comment by mr-c** [comment]
https://github.com/ged-lab/2013-read-aligner/blob/mrc-additions/pairhmm_learning/learn.py now can output JSON directly

**Comment by mr-c** [comment]
@ctb ready for review & merge

File: khmer/__init__.py [hunk 1]
```diff
@@ -164,7 +167,9 @@ def extract_countinghash_info(filename):
 
 
 def calc_expected_collisions(hashtable, force=False, max_false_pos=.2):
-    """Do a quick & dirty expected collision rate calculation on a hashtable.
```
Comment (ctb) [inline khmer/__init__.py]: why this change?

File: tests/test_scripts.py [hunk 2]
```diff
@@ -2207,15 +2207,15 @@ def test_sample_reads_randomly():
 
     seqs = set([r.name for r in screed.open(outfile)])
     print seqs
-    assert seqs == set(['850:2:1:1849:7272/1',  '850:2:1:1849:7272/2',
```
Comment (ctb) [inline tests/test_scripts.py]: why this change?

**Comment by ctb** [comment]
Looks good so far. I'm not sure the test tests anything, does it?  Would be good to have a getter as well as a setter, so that we can have a round-trip test.

Also, we should rope in @luizirber - he's going to be poking at some small-k trained parameters soon.

File: khmer/__init__.py (hunk 1 above)
Comment (mr-c) [inline khmer/__init__.py]: PEP257 fix

File: tests/test_scripts.py (hunk 2 above)
Comment (mr-c) [inline tests/test_scripts.py]: Autopep8.

**Comment by mr-c** [comment]
Will fix #755

**Comment by mr-c** [comment]
- [ ] add getter to C++ & Python APIs

**Comment by mr-c** [comment]
Hmm.. the values for the transition probabilities are getting scrambled. They make it into the C++ array fine but upon creation of the ScoringMatrix something goes wrong..

```
(gdb) p transitions
$7 = {_M_elems = {-0.021973842014145723, -13.73189622448781, 
    -14.869792371737484, -6.0582398155018424, -19.166033822961197, 
    -21.668534163490381, -0.94447280004976863, -1.105331993785005, 
    -7.40360292819022, -6.6908964730845044, -0.39379373931924933, 
    -2.1236734673666091, -7.1043648214967936, -8.8646048755159335, 
    -3.645644436080496, -15.220073662674086, -12.411146320797728, 
    -0.1263680454390087, -8.227232598141855, -11.226627458948142, 
    -2.8013509614037972, -8.078453985883888, -1.9197909720107271, 
    -0.76475134486149254, -2.8525098984653257, -4.5350708169669423, 
    -0.65223888522854956, -2.4570387304176129}}
(gdb) p m_sm.tsc                                                      
$8 = {_M_elems = {0, 0, 0, 0, 0, 0, 5.9445978507618784e-317, 
    2.1219957904712067e-314, 6.9531436082568959e-310, 
    6.9533463822830214e-310, 7.4395772546748228e-317, 
    -nan(0xfffffffffffff), 6.3872609068098017e-317, 
    7.4395851597251562e-317, 0, 8.8931816251424378e-323, 
    7.4395258718476553e-317, 6.9533558078349549e-310, 
    4.9406564584124654e-324, 6.9533459220976308e-310, 
    6.953348481543445e-310, 0, 0, 0, 0, 0, 0, 0}}
(gdb) 
```

**Comment by mr-c** [comment]
Things I tried today: returning the ScoringMatrix as a pointer to avoid compiler generated copying; using a `std::vector` instead of an array; variations on the above.

No success, same erroneous value.

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
Tests pass, thanks @ctb. Coverage dropped a bit, though :-P

**Comment by mr-c** [comment]
@ctb Shall I prepare a PR into your GraphAlign branch?

**Comment by ctb** [comment]
On Fri, May 15, 2015 at 09:37:15AM -0700, Michael R. Crusoe wrote:
> @ctb Shall I prepare a PR into your GraphAlign branch?

Sure!

**Comment by mr-c** [comment]
- [ ] investigate code coverage issue
- [ ] full round trip test
- [ ] after merge to master then merge master into @ctb 's graphalign PR via new PR
- [ ] @luizirber to document out of memory situation in an issue
- [ ] move pairhmm_training script + Makefile into Sandbox
- [ ] move `my_mock.fa` to athyra.ged.msu.edu/mcrusoe

**Comment by mr-c** [comment]
Superseded by #1019
